### ISSUE TITLE
TechDraw: rework annotation tools

### ISSUE THREAD

**Issue by PaddleStroke** [FreeCAD/FPA-grant-proposals#26]
### Proposal description

Currently techdraw has 2 annotation tools. And both are not ideal.
- The first has a very weird UI with + - buttons that you need to use to add lines. Tedious to use. And it does not support any formating.
![image](https://github.com/user-attachments/assets/9a0ac2ee-2ed2-45a4-9aee-1b8a4e1d3b40)

- The second one is better as it let you format text properly. But its UI is still quite clunky.
![image](https://github.com/user-attachments/assets/70125dfb-28b2-4429-86eb-4dcd4aa69aa3)
It has 2 editor widgets, one with HTML capability in a separate dialog. Which can be opened by clicking a button. And you need to click save not close or you loose your text. And after the creation, if you double click the view you can get the HTML editor back. But nothing to change the size of the box

My proposal is to make a new tool replacing existing :

- No task panel. Clicking new annotation makes adds an annotation on the view (as current first tool does)
- When you click and drag on the border of the annotation view, you can resize it
- Right click on the view offers border settings.
- Double clicking (or simply clicking somewhere on the text) starts editing the text. This makes HTML buttons appear and you can edit the text directly. Below is a quick mockup : 
![image](https://github.com/user-attachments/assets/d97a2f53-c151-4150-9ad0-ad353daa17c0)

Fix https://github.com/FreeCAD/FreeCAD/issues/12371

### Deliverables

PR with the feature

### Timeline

Start February. PR in Feb/March. Ends when it merges

### Compensation

Phase 1 :
Development : 5-10 days
PR handling, adjustments, rebasing, fixing conflicts : 3 days

Phase 2 :
Fixing introduced bugs or issues : 2-3 days

Total compensation : 6000€ (after the PR merges).

### About you

Pierre, paddle on the forum, PaddleStroke on github

In the past I have done a lot of work on FreeCAD: https://github.com/FreeCAD/FreeCAD/pulls?q=is%3Apr+author%3APaddleStroke+is%3Aclosed

**Comment by yorikvanhavre** [comment]
If we're going to develop a new text tool, what about trying to unify it with:

* [Draft Text](https://wiki.freecad.org/Draft_Text)
* [Draft ShapeString](https://wiki.freecad.org/Draft_ShapeString)
* https://github.com/FreeCAD/FreeCAD/pull/16928 ( Text workbench)

Don't know how much can be made common, but I would imagine we could have a common UI, and maybe a common "font/text engine" that can be shared between those tools, even if each has a different output

**Comment by PaddleStroke** [comment]
Yes this is a good idea. The HTML formating controls would be a hover-view widget, like the sketcher on-view-parameter spinboxes.
So this could be put in core. This way all workbenches could use it.

Now regarding the techdraw-view object itself, it's quite specific to techdraw. So I do not think things like its resizing and movement can be shared so easily.

**Comment by yorikvanhavre** [comment]
I guess it needs a good definition of what can be made common and what would go in workbench-specific "backends"... 

But I can imagine at some point later on people would want to implement such on-screen stuff like resize, edit... in Draft too, so it might be interesting to have some support for it in a common module

**Comment by prokoudine** [comment]
You'd have to pay a lot of attention to the hover toolbar. GIMP users hate the one they got with quite a passion because it blocks the view of the rest of the page.

**Comment by kadet1090** [comment]
Agree with Aleksandr here, this can quickly become very infuriating. I think that edit should be done in-place so you see how the annotation will look instantly but the edit controls should probably be placed somewhere on top or in the task panel. 

You could also take inspiration from other editors (like word, or conflence) and show the hovered toolbar only if user has both selected something and is hovering over that selection.

It would also be great if such "hovered" toolbar feature would be implemented in a way that makes it possible to use in other workbenches. For example in part design when user has selected edge and is hovering over it we could display small toolbox with stuff like chamfer, fillet etc.

**Comment by chennes** [comment]
Thanks for the submission: the 2025 grant process is going to be slightly different than the 2024. We are switching to a quarterly approval system. That will have two main consequences: first, the submission deadline for Q1 is now March 1, and all submissions will be considered by the committee then (though the larger FreeCAD community will of course continue to discuss the proposal here in this issue), and second, we are going to have a fixed budget for the quarter, so this grant will be competing for a slice of a set funding amount. That budget has not yet been determined, I will let you know here when it has.

**Comment by chennes** [comment]
The FPA has now voted to approve a budget of €20.000 per quarter, for a total of €80.000 in funds to be allocated by the grant program over the course of the year. This value may be revised in Q3 if donations significantly differ from the FPA's financial projections.

**Comment by chennes** [comment]
Thank you again for your grant proposal: it was submitted in time to be evaluated as part of the Q1 2025 grant cycle, and is currently under committee review. In this quarter we received 12 grant requests totaling approximately 61.800 EUR in requests, and we expect to award approximately 15.000 EUR in grants. As you can tell from these numbers we expect the approval process to be highly competitive! We appreciate your participation in the program, and you can expect to hear the results of the  technical review committee's deliberations in two weeks.

**Comment by Reqrefusion** [comment]
I think they call it a floating toolbar, not a hover. It's a bit of a pain to fine-tune its behavior, but it's something I really want to see in FreeCAD. Not just in TechDraw, of course, everywhere.

**Comment by chennes** [comment]
The FPA's vote on the Quarter 1 2025 grant proposals is complete, and this grant was not selected for funding.

Selected comments from reviewers:
* "While it would be really good to have better text tools in FreeCAD, this proposal only addresses a small part of the problem, which turns it a bit expensive for the benefit"
* "It will be nice to have a better GUI for annotations. It is a valuable fix to have, but I feel the compensation is a bit on the high side."
* "A unified text tool would be better than another, different tool"

Overall the challenge with this type of proposal is the opposite of the problem that many other proposals have: rather than being too large in scope, it is too small. The FPA budget does not currently allow for funding developers at a full-time pay rate, so by necessity all FPA funding is expected to form only a part of the money necessary to complete any given project.

**Comment by PaddleStroke** [comment]
I see.
Perhaps the FPA could have considered the +/- 20 PR fixing bugs that I have submitted in just the last month before saying that by funding this it would have been a full-time pay rate.

Bottom line, one has better spend his time polishing their grant requests to make them pretty rather than fixing bugs.
